### ISSUE TITLE
.Net: Add new AudioToText API to use services with locally deployed LLMs

### ISSUE THREAD

**Issue by andreykaone** [microsoft/semantic-kernel#6647]
Thanks a lot to you guys for this project! It gives a very easy way to add AI functionality to already created apps.

But I've noticed that there are Azure and OpenAI connectors for audio-to-text API but no way to use locally deployed LLMs. So it would be nice to have this opportunity to 'talk to' our local models.

So basically, I propose to add a new `OpenAIAudioToTextService(string modelId, Uri endpoint, ..)` constructor with a `Uri` parameter to set which backend to use and an `IKernelBuilder AddOpenAIAudioToText(this IKernelBuilder builder, string modelId, Uri endpoint, ..)` extension for `IKernelBuilder` to configure kernel with STT service.

Creating this issue as [CONTRIBUTING.md](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) suggests discussing these changes with the team.

In fact I already have an implementation for this issue, going to pr it soon

**Comment by markwallace-microsoft** [comment]
Hi @andreykaone thanks for creating the issue and looking forward to seeing your PR. I'm going to assign this to @RogerBarreto so he can help you to get your PR reviewed.

**Comment by dmytrostruk** [comment]
> So basically, I propose to add a new OpenAIAudioToTextService(string modelId, Uri endpoint, ..) constructor with a Uri parameter to set which backend to use and an IKernelBuilder AddOpenAIAudioToText(this IKernelBuilder builder, string modelId, Uri endpoint, ..) extension for IKernelBuilder to configure kernel with STT service.

@andreykaone I would propose for local models to create separate `{Name}AudioToTextService` which implements `IAudioToTextService` interface and add a functionality to access local models. The reason why I wouldn't merge it with OpenAI functionality is because local models could have some features or configuration which won't be applicable to OpenAI and vice versa, OpenAI models may have some functionality which won't be applicable to local models. If we keep functionality for local models in separate service, it will give us more flexibility to extend it in the future.

**Comment by MithrilMan** [comment]
While I see your argument, I'd add that there are some local services that implement OpenAI APIs, so having a way to just pass an URL would help a lot rather than having to create custom implementation of a service.

**Comment by damian-666** [comment]
I see arguments on both sides for this, 
i favor  IAudioToTextService

TLDR
you what the current state of affairs is and how I'm managing to deal with it, given that my right arm stopped working last week because of disk. broken disk in my neck.  



I don't use OpenAI so called plug in things. I don't trust them. I'm not gonna  send my data to some url, and I don't really want to set up my own local server either, cause that's  weird..  

so i use windows VoiceAcess  (buggy) but WIP and offline 
plus this..does  require a URL.
I should make a clip board listener like power tools has but thats not ideal

https://platform.openai.com/docs/guides/speech-to-text/improving-reliability

so i use this to use chatgpt 
https://github.com/wieslawsoltes/ChatGPT


**Raw text from voice access  dev insider, latest : Voice Access**  -->Hey, let me just chime in on this nut I haven't used semantic kernel at all. I have no experience with it, but I am able to  I've had a session with Chad GPT to ask me exactly how to make voice access work and right now my right hand does not type, so I need it to work.  

I can just give you a prompt that I use that does. fix. The voice access that's already in windows. And that I'm seeing that one of the reasons that it's kind of long is that they are using context to fix. the input, but this kind of work is going into voice access slowly and I'm losing. It's not quite finished. I just saw a **switcher** word. So maybe they're already working on this cause switch your words should have been changed. That's not switcher word isn't even a word. So this is, this is raw. This is coming from voice to text. Now I'm going to send it through my LLM, and I'm going to give you the prompt.
I'm just giving you a use case here.   <---- Raw text from voice access  dev insider  build lastest WIP


--prompt
You are a helpful assistant named TypoFixerDictionationOrMenuPicker.  for a long input,  just fix typos in my text, make no major grammar or wording changes or reinterpretations.  on a second pass, if it appears  that  there is  the wrong word or acronym that's out of  context you can **switch** the word,  
Output the fixed text, then
give a numbered list of the changed typos, then the changed words on the second pass.    For acronyms if they don't fix in context include those..    contains  a  list like     [ Open File, File Close, Open Last Workspace],   just choose the closest match    for the word or words following the list.  then output that as a "Change Context and Command is"   , then give the likely match



--->llm fix
Hey, let me just chime in on this because I haven't used Semantic Kernel at all. I have no experience with it, but I have had a session with ChatGPT to figure out exactly how to make voice access work. Right now, my right hand does not type, so I need it to work. 

I can give you a prompt that I use to fix the voice access that's already in Windows. One of the reasons it's kind of long is that it's using context to fix the input, but this kind of work is going into voice access slowly. I'm losing some accuracy because it's not quite finished. I just saw it **switch** a word. They might already be working on this because "switcher word" isn't even a word. This is raw; this is coming from voice to text. 

Now, I'm going to send it through my LLM (Language Model) and give you the prompt. I'm just giving you a use case here.

### Changed Typos:
1. and right now -> Right now,
2. and that I'm seeing -> I'm seeing 
3. context to fix. the input -> context to fix the input
4. as switch your words -> I as an switch your words; this is bad.  <--??  
5. so this is -> so, this is

### Changed Words:
None detected.

### Change Context and Command is:
None detected.

![clippy (Phone)](https://github.com/user-attachments/assets/2be775e4-d789-46b1-8744-535681a24196)

**Comment by dmytrostruk** [comment]
> While I see your argument, I'd add that there are some local services that implement OpenAI APIs, so having a way to just pass an URL would help a lot rather than having to create custom implementation of a service.

The main question is - how easy these local services could adapt to OpenAI API changes if there will be any in the future? Let's say, if OpenAI will deprecate some configuration field, it will be deprecated in OpenAI SDK as well. What if this field will still be applicable in local services? How to evaluate which AI providers and local models will be impacted of such change? 

I agree that re-using already existing functionality by simply passing another URL is great, it saves a lot of time in implementation and maintenance. I'm pretty sure that it will cover some basic scenarios without any problems and users can leverage that. But I think if either OpenAI or AI provider with local models will need some updates or extensions that won't be applicable by other side - this will be problematic. That's why I'm not sure if re-using OpenAI SDK for local models provided by other AI providers is a good choice in the long term.

**Comment by MithrilMan** [comment]
> > While I see your argument, I'd add that there are some local services that implement OpenAI APIs, so having a way to just pass an URL would help a lot rather than having to create custom implementation of a service.
> 
> The main question is - how easy these local services could adapt to OpenAI API changes if there will be any in the future? Let's say, if OpenAI will deprecate some configuration field, it will be deprecated in OpenAI SDK as well. What if this field will still be applicable in local services? How to evaluate which AI providers and local models will be impacted of such change?
> 
> I agree that re-using already existing functionality by simply passing another URL is great, it saves a lot of time in implementation and maintenance. I'm pretty sure that it will cover some basic scenarios without any problems and users can leverage that. But I think if either OpenAI or AI provider with local models will need some updates or extensions that won't be applicable by other side - this will be problematic. That's why I'm not sure if re-using OpenAI SDK for local models provided by other AI providers is a good choice in the long term.

I've to say I missread the topic because I was looking for an implementation of TextToAudio, but it does suffer from the same problem so this issue is still relevant.

I don't think you should worry if a 3rd party project follows or not the OpenAI API, as long the SK Connector has a clear notice, like: "we support OpenAI compatible APIs but don't support specific implementations so if a 3rd party service doesn't adhere to current OpenAI implementation it could break compatibility"

I took a look at your implementation and it could have been easily fixed if there was a way to register the service passing an Endpoint, but the only connector that do that is Azure and its implementation is sealed and doesn't have any extension point.
I can understand the desire to push for Azure (a way to monetize for MS who is backing the development) if this is the intention, in any case leaving open the way to specify an endpoint for every service could be an easy way to have a ready-to-go implementation for every service.

I think this is valid for any major generative AI service around (text, image, audio, etc...): as long as some establish itself, some opensource alternative pop out parroting the same API interface for an easy swap.

At the end leaving users to customize endpoints is a no-brainer/no-cost implementation for you (you don't even need to support it in a specific way)

P.S.
Text-To-Audio should support streaming too :)

**Comment by dmytrostruk** [comment]
> I don't think you should worry if a 3rd party project follows or not the OpenAI API, as long the SK Connector has a clear notice, like: "we support OpenAI compatible APIs but don't support specific implementations so if a 3rd party service doesn't adhere to current OpenAI implementation it could break compatibility"

That's why we should be careful with this approach because at some point it could break compatibility, which is the scenario we probably want to avoid :) I have to mention that there is no team final decision about this functionality yet, I'm just sharing my personal thoughts. This topic will be discussed within internal team, we will probably create some ADR documentation to capture pros and cons of each approach etc.

> I took a look at your implementation and it could have been easily fixed if there was a way to register the service passing an Endpoint, but the only connector that do that is Azure and its implementation is sealed and doesn't have any extension point.
I can understand the desire to push for Azure (a way to monetize for MS who is backing the development) if this is the intention, in any case leaving open the way to specify an endpoint for every service could be an easy way to have a ready-to-go implementation for every service.

I'm not sure it's the reason, because both Azure and OpenAI connectors are using the same underlying SDK at the moment. There is on-going work to split these connectors by separate SDKs, it's in progress, but as for now, in `main` branch both services use the same SDK. I think the main reason why it's exposed in Azure connector is because when you deploy the model in your environment, it gets unique URL which contains your deployment name, so there should be a way how to specify the endpoint, while in case of OpenAI the endpoint is constant. Also, for chat completion there is already overload that accepts custom endpoint for OpenAI service:
https://github.com/microsoft/semantic-kernel/blob/12a877b8297bfb39987843b3f95acf606cb163ca/dotnet/src/Connectors/Connectors.OpenAI/OpenAIServiceCollectionExtensions.cs#L1090-L1093

The reason why `endpoint` parameter wasn't added to audio-related method is because the case for audio with local models wasn't considered during the implementation phase.

By the way, each extension method accepts `HttpClient`, so in theory it should be possible to implement custom delegating handler, that will change base address inside `SendAsync` method to the custom one:
https://github.com/microsoft/semantic-kernel/blob/12a877b8297bfb39987843b3f95acf606cb163ca/dotnet/src/Connectors/Connectors.OpenAI/OpenAIServiceCollectionExtensions.cs#L1646-L1652

If you could check this approach and share your results with us, that would be awesome!

**Comment by MithrilMan** [comment]
> That's why we should be careful with this approach because at some point it could break compatibility, which is the scenario we probably want to avoid :) I have to mention that there is no team final decision about this functionality yet, I'm just sharing my personal thoughts. This topic will be discussed within internal team, we will probably create some ADR documentation to capture pros and cons of each approach etc.

good
 
> The reason why `endpoint` parameter wasn't added to audio-related method is because the case for audio with local models wasn't considered during the implementation phase.
> 
> By the way, each extension method accepts `HttpClient`, so in theory it should be possible to implement custom delegating handler, that will change base address inside `SendAsync` method to the custom one:
> 
> https://github.com/microsoft/semantic-kernel/blob/12a877b8297bfb39987843b3f95acf606cb163ca/dotnet/src/Connectors/Connectors.OpenAI/OpenAIServiceCollectionExtensions.cs#L1646-L1652
> 
> If you could check this approach and share your results with us, that would be awesome!

this is one of the first thing I tried to do but unfortunately in my code I don't use a kernel builder but I use 


```
public static IServiceCollection AddOpenAI(this IServiceCollection services, IConfiguration configuration, VoiceGenerationOptions voiceGenerationOptions)
{
   var providerName = OpenAIOptions.PROVIDER_NAME;

   if (voiceGenerationOptions.DefaultProvider != providerName) return services; //TODO: Add support for multiple providers
   if (voiceGenerationOptions.Providers.ContainsKey(OpenAIOptions.PROVIDER_NAME) == false) return services;

   var settings = configuration.GetSection(OpenAIOptions.SECTION_NAME).Get<OpenAIOptions>()
               ?? throw new InvalidOperationException("OpenAI options are not specified");

   services.Configure<OpenAIOptions>(configuration.GetSection(OpenAIOptions.SECTION_NAME));
   services.TryAddKeyedTransient<ISpeechGenerator, OpenAiSpeechGenerator>(providerName);
   services.AddOpenAITextToAudio(
      modelId: settings.Model,
      apiKey: configuration.GetOpenAIKey(),
      serviceId: providerName);

   return services;
}
```

This is a snippet of my code, I use `AddOpenAITextToAudio` that doesn't have an overload with HttpClient.
Later I'll try implementing a custom extension that use the constructor with the httpclient but the DI service registration shouldn't be something user should do in this scenario because it's an internal detail of SK.

Probably something along this line (here I pass an endpoint but could pass an httpclient) could be enough

```
public static IServiceCollection AddOpenedAITextToAudio(
    this IServiceCollection services,
    string modelId,
    Uri endpoint,
    string? serviceId = null)
{
   ArgumentNullException.ThrowIfNull(services);
   ArgumentException.ThrowIfNullOrWhiteSpace(modelId);
   ArgumentNullException.ThrowIfNull(endpoint);


   return services.AddKeyedSingleton<ITextToAudioService>(serviceId, (serviceProvider, _) =>
   {
      var httpClient = (serviceProvider?.GetService<HttpClient>())
         ?? throw new InvalidOperationException("HttpClient is not registered");

      httpClient.BaseAddress = endpoint;

      return new OpenAITextToAudioService(
         modelId: modelId,
         apiKey: "FAKE-KEY",
         organization: null,
         httpClient: httpClient,
         loggerFactory: serviceProvider.GetService<ILoggerFactory>());
   });
}
```

**Comment by RogerBarreto** [comment]
@MithrilMan, with our updated OpenAI V2 Connector already in `main`, let me know if you still need this specific `endpoint` requirement.

Currently you are new able to do so providing your own `OpenAIClient` + `Endpoint` into one of our `KernelBuilder` or `ServiceCollection` extension methods which accept the `openAIClient`.

**Comment by RogerBarreto** [comment]
Closing as no feedback was provided, assuming the suggested approach is the one moving forward.
